File: core-swift/Sources/HanguljiCore/JamoComposer.swift
// Sources/HanguljiCore/JamoComposer.swift

public struct Syllable: Equatable {
    public var initial: Consonant?
    public var vowel: Vowel?
    public var final: Consonant?

    public init(initial: Consonant? = nil, vowel: Vowel? = nil, final: Consonant? = nil) {
        self.initial = initial
        self.vowel = vowel
        self.final = final
    }

    /// 종성 인덱스 (유니코드 한글 조합 공식용). ㄸㅃㅉ은 종성 불가 → nil
    private static let jongseongIndex: [Consonant: Int] = [
        .g: 1, .gg: 2, .n: 4, .d: 7, .r: 8, .m: 16, .b: 17, .s: 19, .ss: 20,
        .ng: 21, .j: 22, .ch: 23, .k: 24, .t: 25, .p: 26, .h: 27,
    ]

    static func canBeFinal(_ c: Consonant) -> Bool { jongseongIndex[c] != nil }

    /// 화면 표시용 한글 (완성형 블록, 미완성이면 자모 낱자)
    public var hangul: String {
        if let i = initial, let v = vowel {
            let ci = Consonant.allCases.firstIndex(of: i)!
            let vi = Vowel.allCases.firstIndex(of: v)!
            let fi = final.flatMap { Self.jongseongIndex[$0] } ?? 0
            let code = 0xAC00 + (ci * 21 + vi) * 28 + fi
            return String(UnicodeScalar(code)!)
        }
        if let i = initial { return String(i.rawValue) }
        if let v = vowel { return String(v.rawValue) }
        return ""
    }
}

public struct JamoComposer {
    public private(set) var jamos: [Jamo] = []

    public init() {}

    public var isEmpty: Bool { jamos.isEmpty }

    public mutating func append(_ jamo: Jamo) { jamos.append(jamo) }

    @discardableResult
    public mutating func backspace() -> Bool {
        guard !jamos.isEmpty else { return false }
        jamos.removeLast()
        return true
    }

    public mutating func clear() { jamos.removeAll() }

    /// 모음 조합 (일본어 표기에 필요한 것 + 표준 몇 개)
    private static let vowelCombinations: [Vowel: [Vowel: Vowel]] = [
        .o: [.a: .wa, .ae: .wae, .i: .oe],
        .u: [.eo: .wo, .e: .we, .i: .wi],
        .eu: [.i: .ui],
    ]

    /// 자모 스트림 전체를 다시 조합 (스트림이 짧으므로 매번 재계산)
    public var syllables: [Syllable] {
        var result: [Syllable] = []
        var cur = Syllable()

        func flush() {
            if cur != Syllable() { result.append(cur) }
            cur = Syllable()
        }

        for jamo in jamos {
            switch jamo {
            case .consonant(let c):
                if cur.vowel == nil {
                    if cur.initial == nil {
                        cur.initial = c
                    } else {
                        flush()
                        cur.initial = c
                    }
                } else if cur.initial != nil, cur.final == nil, Syllable.canBeFinal(c) {
                    cur.final = c
                } else {
                    flush()
                    cur.initial = c
                }
            case .vowel(let v):
                if let f = cur.final {
                    // 받침 재해석: 받침을 다음 음절 초성으로
                    cur.final = nil
                    flush()
                    cur.initial = f
                    cur.vowel = v
                } else if cur.vowel == nil {
                    cur.vowel = v
                } else if let combined = Self.vowelCombinations[cur.vowel!]?[v] {
                    cur.vowel = combined
                } else {
                    flush()
                    cur.vowel = v
                }
            }
        }
        flush()
        return result
    }
}

File: android/app/src/main/kotlin/com/mastergear/hangulji/core/JamoComposer.kt
package com.mastergear.hangulji.core

data class Syllable(
    val initial: Consonant? = null,
    val vowel: Vowel? = null,
    val final: Consonant? = null,
) {
    /** 화면 표시용 한글 — SPEC §2.7 유니코드 완성형 조합 공식 (미완성이면 낱자) */
    val hangul: String
        get() {
            if (initial != null && vowel != null) {
                val fi = final?.let { jongseongIndex[it] } ?: 0
                val code = 0xAC00 + (initial.ordinal * 21 + vowel.ordinal) * 28 + fi
                return String(Character.toChars(code))
            }
            if (initial != null) return initial.char.toString()
            if (vowel != null) return vowel.char.toString()
            return ""
        }

    companion object {
        /** 종성 인덱스 (SPEC §2.7). ㄸㅃㅉ 없음 = 종성 불가 (§2.5) */
        private val jongseongIndex: Map<Consonant, Int> = mapOf(
            Consonant.G to 1, Consonant.GG to 2, Consonant.N to 4, Consonant.D to 7,
            Consonant.R to 8, Consonant.M to 16, Consonant.B to 17, Consonant.S to 19,
            Consonant.SS to 20, Consonant.NG to 21, Consonant.J to 22, Consonant.CH to 23,
            Consonant.K to 24, Consonant.T to 25, Consonant.P to 26, Consonant.H to 27,
        )

        fun canBeFinal(c: Consonant): Boolean = jongseongIndex.containsKey(c)
    }
}

class JamoComposer {
    private val jamos = mutableListOf<Jamo>()

    val isEmpty: Boolean get() = jamos.isEmpty()

    fun append(jamo: Jamo) { jamos.add(jamo) }

    fun backspace(): Boolean {
        if (jamos.isEmpty()) return false
        jamos.removeAt(jamos.size - 1)
        return true
    }

    fun clear() { jamos.clear() }

    /** 모음 조합 결합표 (SPEC §2.4) */
    private val vowelCombinations: Map<Vowel, Map<Vowel, Vowel>> = mapOf(
        Vowel.O to mapOf(Vowel.A to Vowel.WA, Vowel.AE to Vowel.WAE, Vowel.I to Vowel.OE),
        Vowel.U to mapOf(Vowel.EO to Vowel.WO, Vowel.E to Vowel.WE, Vowel.I to Vowel.WI),
        Vowel.EU to mapOf(Vowel.I to Vowel.UI),
    )

    /** 자모 스트림 전체를 매번 처음부터 재조합 (SPEC §2.1 — 스트림이 짧아 단순함 우선) */
    val syllables: List<Syllable>
        get() {
            val result = mutableListOf<Syllable>()
            var cur = Syllable()

            fun flush() {
                if (cur != Syllable()) result.add(cur)
                cur = Syllable()
            }

            for (jamo in jamos) when (jamo) {
                is Jamo.C -> {   // SPEC §2.2
                    val c = jamo.consonant
                    if (cur.vowel == null) {
                        if (cur.initial == null) {
                            cur = cur.copy(initial = c)
                        } else {
                            flush()
                            cur = Syllable(initial = c)
                        }
                    } else if (cur.initial != null && cur.final == null && Syllable.canBeFinal(c)) {
                        cur = cur.copy(final = c)
                    } else {
                        flush()
                        cur = Syllable(initial = c)
                    }
                }
                is Jamo.V -> {   // SPEC §2.3
                    val v = jamo.vowel
                    val f = cur.final
                    if (f != null) {
                        cur = cur.copy(final = null)   // 받침 재해석: 종성을 다음 음절 초성으로
                        flush()
                        cur = Syllable(initial = f, vowel = v)
                    } else if (cur.vowel == null) {
                        cur = cur.copy(vowel = v)
                    } else {
                        val combined = vowelCombinations[cur.vowel]?.get(v)
                        if (combined != null) {
                            cur = cur.copy(vowel = combined)
                        } else {
                            flush()
                            cur = Syllable(vowel = v)
                        }
                    }
                }
            }
            flush()
            return result
        }
}
